File: Binary_Search/911.Online_Election/Python/911.Online_Election.py
class TopVotedCandidate(object):
    def __init__(self, persons, times):
        """
        :type persons: List[int]
        :type times: List[int]
        """

        ############
        #Initialize
        ##### Record , Lead arrays #####
        self.lead_arry, self.record_arry = [persons[0]], times[:]

        ####################
        #Whole process/flow
        record_dict = {persons[0]: 1} #Record dictionary/hashmap

        record_max_val = 1 #Record maximun value

        total_length = len(persons) #Total length

        for persons_idx in range(1, total_length):
            (self.lead_arry).append((self.lead_arry)[(-1)]) #Keep updating/recording

            ##### Check if the current person-id existed or not #####
            if (persons[persons_idx] not in record_dict):
                record_dict[persons[persons_idx]] = 1 #Keep updating/recording
            else:
                record_dict[persons[persons_idx]] += 1 #Keep updating/recording

            ##### Check if the current summary counter matched conditions or not #####
            if (record_dict[persons[persons_idx]] >= record_max_val):
                record_max_val = record_dict[persons[persons_idx]] #Keep updating/overwriting

                (self.lead_arry)[(-1)] = persons[persons_idx] #Keep updating/overwriting
            else:
                pass


    def q(self, t):
        """
        :type t: int
        :rtype: int
        """

        ############
        #Initialize
        ##### Record pointers (i.e. left, right) #####
        record_left_ptr, record_right_ptr = 0, (len(self.record_arry) - 1)

        ####################
        #Whole process/flow
        while (record_left_ptr < record_right_ptr):
            record_mid_ptr = ((record_left_ptr + record_right_ptr + 1) // 2) #Record middle pointer

            ##### Check if the current times boundary matched conditions or not #####
            if ((self.record_arry)[record_mid_ptr] > t):
                record_right_ptr = (record_mid_ptr - 1) #Keep updating/overwriting
            else:
                record_left_ptr = record_mid_ptr #Keep updating/overwriting

        return (self.lead_arry)[record_left_ptr]


# Your TopVotedCandidate object will be instantiated and called as such:
# obj = TopVotedCandidate(persons, times)
# param_1 = obj.q(t)

File: Binary_Search/911.Online_Election/C++/911.Online_Election.cpp
class TopVotedCandidate 
{
    /*Initialize*/
    ///// Record, Lead vectors /////
    vector<int> record_vect, lead_vect;

public:
    TopVotedCandidate(vector<int> & persons, vector<int> & times) 
    {
        //=======================================//
        // Input type:                           //
        //   - vector<int> ref. (persons vector) //
        //   - vector<int> ref. (times vector)   //
        // Return type:                          //
        //   - void (no return)                  //
        //=======================================//

        /*Initialize*/
        ///// Total length /////
        int total_length = persons.size();

        ///// Record dictionary, hashmap /////
        unordered_map<int, int> record_map;

        ///// Record maximun summary counter /////
        int record_max_cnter;


        /*Whole process, flow*/
        (this->lead_vect).emplace_back(persons[0]); //Update, Record

        ((record_map[persons[0]])++); //Update, Record

        record_max_cnter = 1; //Update, Overwrite

        for (int persons_idx = 1; (persons_idx < total_length); (persons_idx++)) //Whole
        {
            (this->lead_vect).emplace_back((this->lead_vect).back()); //Keep updating, recording

            ((record_map[persons[persons_idx]])++); //Keep updating, recording

            ///// Check if the current summary counter mathced conditions or not /////
            if (record_map[persons[persons_idx]] >= record_max_cnter)
            {
                record_max_cnter = record_map[persons[persons_idx]]; //Keep updating, overwriting

                (this->lead_vect).back() = persons[persons_idx]; //Keep updating, overwriting
            }
            else { ; }

        } //Whole

        (this->record_vect) = vector<int>(times.begin(), times.end()); //Keep updating, overwriting
    }

    ~TopVotedCandidate()
    {
        //======================//
        // Input type:          //
        //   - void (no input)  //
        // Return type:         //
        //   - void (no return) //
        //======================//
        ;
    }
    
    int q(int t) 
    {
        //=========================//
        // Input type:             //
        //   - int (t-times value) //
        // Return type:            //
        //   - int (result id)     //
        //=========================//

        /*Initialize*/
        ///// Record pointers (i.e. left, right) /////
        int record_left_ptr = 0, record_right_ptr = ((this->record_vect).size() - 1);

        /*Whole process, flow*/
        while (record_left_ptr < record_right_ptr) //Binary-search
        {
            int record_mid_ptr = ((record_left_ptr + record_right_ptr + 1) / 2); //Record middle pointer

            ///// Check if the current times matched conditions or not /////
            if ((this->record_vect)[record_mid_ptr] > t) { record_right_ptr = (record_mid_ptr - 1); } //Keep updating, overwriting
            else { record_left_ptr = record_mid_ptr; } //Keep updating, overwriting

        } //Binary-search

        return (this->lead_vect)[record_left_ptr];
    }
};

/**
 * Your TopVotedCandidate object will be instantiated and called as such:
 * TopVotedCandidate* obj = new TopVotedCandidate(persons, times);
 * int param_1 = obj->q(t);
 */
